### PR TITLE
[Bugfix] Fix sequence-parallel detection for small decode batches

### LINKED ISSUE #50154: [Bug]: GLM 5.2; Garbled output with sequence-parallel MoE on small decode batches (TP+EP+DP)
> [!NOTE]
> This issue and the suggested fix were drafted with the assistance of Claude Opus 4.6.

### Your current environment

```
vllm-0.26.1rc1.dev18+gd223c900d (nightly, also reproduced on 0.23.1rc1.dev1403)
8x NVIDIA B200, 2 nodes
```

### Model

`zai-org/GLM-5.2-FP8` (744B MoE, 256 experts)

### Parallel config

TP=4, DP=2, EP enabled (`--data-parallel-size-local=1`, `--data-parallel-hybrid-lb`)

### How to reproduce

```bash
vllm serve zai-org/GLM-5.2-FP8 \
  --tensor-parallel-size 4 \
  --enable-expert-parallel \
  --data-parallel-size 2 \
  --data-parallel-size-local 1 \
  --data-parallel-hybrid-lb \
  --trust-remote-code \
  --gpu-memory-utilization 0.90 \
  --max-model-len 131072 \
  --max-num-seqs 32 \
  --enforce-eager \
  --async-scheduling \
  --enable-flashinfer-autotune \
  --no-enable-prefix-caching
```

Send any short prompt:
```bash
curl http://localhost:8000/v1/completions \
  -d '{"model":"zai-org/GLM-5.2-FP8","prompt":"The capital of France is","max_tokens":30,"temperature":0}'
```

### Raw garbled output

```json
{
  "text": " Paris. Paris Paris Paris Paris Paris the the most. Paris and Paris a Paris1140 and41.1,1.0.",
  "system_fingerprint": "vllm-0.26.1rc1.dev18+gd223c900d-tp4-dp2-ep-a53f56b6"
}
```

Another sample (from 0.23.1):
```json
{
  "text": " Canberra country country country country USA Paris Paris Paris Paris Paris Paris Paris Paris Germany Paris Germany Japan Germany Germany Germany Germany Germany..1 Japan Germany Germany.1......1....................1... the the",
  "system_fingerprint": "vllm-0.23.1rc1.dev1403+g4080263bb-tp4-dp2-ep-0d49e890"
}
```

### Root cause

In `vllm/model_executor/models/deepseek_v2.py`, `input_is_sequence_parallel` is detected by:

```python
input_is_sequence_parallel = (
    self.use_sequence_parallel_moe
    and residual is not None
    and hidden_states.shape[0] != full_num_tokens
)
```

For small decode batches where `full_num_tokens < tp_world_size` (e.g. 1-3 tokens with TP=4), the preceding `reduce_scatter` produces `hidden_states` of the same size as `full_num_tokens` (since `ceil(n/tp) == n` for `n < tp`). This makes the condition `False` even when SP is active.

Downstream consequence: `residual` is not chunked (line 1332), so `hidden_states` (now scattered) and `residual` (still full-size) are misaligned when passed to the next layer. This accumulates garbage values across layers, producing garbled output.

### Observations

- Reproduces with `--enforce-eager` on both v0.23.1 and v0.26.1 nightly.
- Does NOT reproduce with `cudagraph_mode=full` (auto-downgraded to `FULL_AND_PIECEWISE`) on v0.26.1. The exact mechanism by which CUDA graph mode avoids this bug is not yet determined.
- Triggers during decode phase whenever a DP rank has fewer active sequences than `tp_world_size` (e.g. < 4 concurrent requests on one rank with TP=4).
- Affects any MoE model using `use_sequence_parallel_moe=True` with TP+EP+DP and small batch sizes.

### Suggested fix

```python
# In vllm/model_executor/models/deepseek_v2.py, DeepseekV2DecoderLayer.forward()

# Fix 1: detection — also catch the case where full_num_tokens < tp_world_size
input_is_sequence_parallel = (
    self.use_sequence_parallel_moe
    and residual is not None
    and (hidden_states.shape[0] != full_num_tokens
         or full_num_tokens < get_tensor_model_parallel_world_size())
)

# Fix 2: residual chunking — use shape comparison instead of boolean flag
# (around line 1332)
if residual.shape[0] != hidden_states.shape[0]:
    residual = sequence_parallel_chunk(residual)
```

PR: #50155

### PR DESCRIPTION
**Pull request by ZhanqiuHu**
> [!NOTE]
> This PR was drafted with the assistance of Claude Opus 4.6.

## Fix

Fixes #50154

### Problem

In `DeepseekV2DecoderLayer.forward()`, `input_is_sequence_parallel` is determined by:

```python
input_is_sequence_parallel = (
    self.use_sequence_parallel_moe
    and residual is not None
    and hidden_states.shape[0] != full_num_tokens
)
```

For small decode batches where `full_num_tokens < tp_world_size` (e.g. 1-3 tokens with TP=4), `reduce_scatter` produces `hidden_states` with the same size as `full_num_tokens` (since `ceil(n/4) == n` when `n < 4`). This makes the shape comparison `False` even when SP is active, so `residual` is never chunked — causing shape misalignment that accumulates garbage across layers.

### Fix

1. Add fallback: also detect SP when `full_num_tokens < tp_world_size`, since scatter is effectively a no-op in this case but SP state is still active.
2. Use `residual.shape[0] != hidden_states.shape[0]` for the residual chunking decision — more robust than relying on the boolean flag alone.

### Testing

- Aggregated deployment: TP=4, EP=4, DP=2, GLM-5.2-FP8 (744B), `--enforce-eager`
- Without fix: garbled output on all short prompts
- With fix: correct generation, gsm8k accuracy ~0.95

### Raw output comparison

**Before (garbled):**
```
"Paris. Paris Paris Paris Paris Paris the the most. Paris and Paris a Paris1140 and41.1,1.0."
```

**After (correct):**
```
"Paris. It is the largest city in France and the country's political, economic, cultural, and commercial center."
```

### CODE REVIEW

File: vllm/model_executor/models/deepseek_v2.py [hunk 1]
```diff
@@ -1334,7 +1330,7 @@ def forward(
             # pad if not divisible by world size
             hidden_states = torch.nn.functional.pad(hidden_states, (0, 0, 0, sp_pad))
             hidden_states = tensor_model_parallel_reduce_scatter(hidden_states, 0)
-            if not input_is_sequence_parallel:
+            if residual.shape[0] != hidden_states.shape[0]:
                 residual = sequence_parallel_chunk(residual)
```
Comment (mgoin) [inline vllm/model_executor/models/deepseek_v2.py:1346]: Any reason not to put this new condition into input_is_sequence_parallel's definition above and keep this as before?

Comment (ZhanqiuHu) [reply vllm/model_executor/models/deepseek_v2.py:1346]: Good point. Fixed it.
